### ISSUE TITLE
Errata

### ISSUE THREAD

**Issue by asdfjkl** [asdfjkl/neural_network_chess#1]
If you spot mistakes, please leave a note as a reply to this issue.

**Comment by asdfjkl** [comment]
page 54: Figure 2.15: Baikinman, appylied sobel operator and thresholding -> applied

**Comment by asdfjkl** [comment]
page 219: Figure 4.23: White to move. There is no reasonable alternative to Bb3 that any
chessplayer would play.. (double colon at the end)

**Comment by MSoszynski** [comment]
p7

"This book is brief"
> This book is a brief

"romantic"
> Romantic

"referring"
> Referring

p8

"a chess players perspective"
> a chess player's perspective

"44... Rd1"
This way to present Black-first moves isn't consistent throughout the book. Compare the very next line on the same page or "1....Bxd1?" on p119. Most books use 3 dots (without a space) not four.

p9

"According to Hsiu"
> According to Hsu

p11

"and Ananad,"
> and Anand,

p12

"computer scientists perspective,"
> computer scientist's perspective,

p13

"Based these evaluations,"
> Based on these evaluations,

p14

"•find a number of candidate moves. Then for each of them sequentially
•calculate all subsequent positions and variation trees that result of those
candidate moves as deep as possible"
> •find a number of candidate moves. Then for each of them sequentially...
•calculate all subsequent positions and variation trees that result from those
candidate moves as deep as possible

"focus on one specific lines, then decide by instinct
instead by rational though"
> focus on one specific line, then decide by instinct instead of by rational thought

p15

"have computers plagued"
> have plagued computers

"to processes and judge"
> to process and judge

p16

"But how do become good player so good at chess?"
> But how do good players become so good at chess?

I hope that's helpful.

Marek Soszynski

**Comment by derestup** [comment]
page 31			dnet1/dw1 = ... = 1	
page 32			... create[s] a small bias.
page 49			... a few lines [of] our simple ...
page 55			equation 2.8 [but your equations are not numbered!]
page 59			Figure 2.17: The third filter matrix of applied to the image of Baikinman [misformed caption]
page 61			... we weigh[t] ...
page 69			... medi[c]al ...
page 69			... it's an[d] old joke ...
page 77			... pawns on the seventh resp. first [second?] rank.
page 89			... our we hit our ...
pages 96, 206, 228	silicon[e]
page 125		Another real achievement ... [misformed sentence]
page 127		... under[s]tand ...
page 132		... exemplary[ly] ...
page 147		... lilkely ...
page 150		... Blacks or White's ...
page 159		... only if ... [will] the new network ...
page 162		not "Expect maybe ..." but "Except maybe ..."
page 175		player[']s
page 176		cons[e]quently
page 191		pro[g]ramming
page 192		... state of the art [of] 
page 217 		under[s]tand

And thank you very much for the book!

**Comment by asdfjkl** [comment]
(all fixed in v1.2)

**Comment by jkrabbenbos** [comment]
Page 12, 14 Hendrik's = Hendriks' (Willie's surname is Hendriks, see also https://grammar.yourdictionary.com/punctuation/apostrophe-rules.html for more info on the use of the apostrophe)
Page 12 Hendrik = Hendriks
In the bibliography the name is written correct!

page 18 HexapwanZero = HexapawnZero

Very interesting and informative book on the use of neural nets in Computer Chess.

**Comment by asdfjkl** [comment]
page 256: instead we desire to select the index 0 only in about 10 percent of
all cases, index 0 in about 20 percent of all cases and index 2 in about 70 percent
of all cases

should read: index 0, ... index 1, ... index 2

**Comment by asdfjkl** [comment]
page 262: it’s that alpha-beta searcher[s]! will prevail
for chess.

**Comment by asdfjkl** [comment]
(all fixed in v1.3)

**Comment by d-kraft** [comment]
page 71 line 3
instead of "the the"
write "take the"

**Comment by d-kraft** [comment]
page 62 last line of paragraph three
I would prefer a comma instead of a hyphen after tensorflow, or -- but worse -- a dash (two hyphens in LaTeX)

**Comment by asdfjkl** [comment]
page 28: should be (0.95 − 0)^2 instead of (0.95 − 1)^2

**Comment by asdfjkl** [comment]
page 30: global should be total

**Comment by asdfjkl** [comment]
page 79: of tactical threads. -> threats

**Comment by asdfjkl** [comment]
page 97: If you do not implement chess knowledge in the implementation function -> evaluation function

**Comment by asdfjkl** [comment]
page 99: 37.Be4 (space!)

**Comment by asdfjkl** [comment]
p148 In contrary, the rather universal gradient policy rein-
forcement learning is _used_ to improve the SL policy network

**Comment by asdfjkl** [comment]
p212: agrees while also searching with around 7,500,000 nodes per
second _and_ evaluates this at

**Comment by d-kraft** [comment]
p. 65 par. 4 line 3

could **be** that

**Comment by d-kraft** [comment]
p. 69 4th line above bottom

common themes

**Comment by d-kraft** [comment]
p. 70 third line from bottom

all required network elements, **there.**

**Comment by d-kraft** [comment]
p. 73 par. 3 line 2

none instead of noone? am not sure

**Comment by d-kraft** [comment]
p. 96 line 2

w.r.t. to 
the to is redundant

**Comment by AlexisNizard** [comment]
p. 155 ,the two formulas that give 1.2 and 1.0 as the result use wrong values (not the values used in Figure 4.8). 

The first formula should be 0.6 * (sqrt(1+1)/2) = 0.42 = 0.5+0.42 = 0.92 
And second should be 0.7 * (sqrt(1+1)/2) = 0.4 + 0.49 = 0.89 ,according to the values in Figure 4.8.

0.92 is still greater than 0.89 so the reasoning doesn't change

**Comment by asdfjkl** [comment]
(all fixed up to here), #todo: release new draft

**Comment by jbscj** [comment]
P204 & P205
Formula 4.1 & Formula 4.2: It seems that the subscripts and superscripts of the terms in the first column should be swapped for consistency
P206 has 3 additional formulae within the text with the above problem
P207 & P208 has formulae 4.3 and 4.4 with the above problem
P209 b1 -> b_1

**Comment by shindavid** [comment]
Everywhere else, as far as I can tell, the "E" in the acronym NNUE stands for the word "efficiently" ([example](https://en.wikipedia.org/wiki/Efficiently_updatable_neural_network)). But in this book, the word "effectively" is used.

**Comment by asdfjkl** [comment]
(all fixed in v1.5)

**Comment by asdfjkl** [comment]
page 185, "will *be*": "These games will naturally of very poor quality initially and it wi"

**Comment by asdfjkl** [comment]
p185: ... rethink it   ... instead of ... summarize what they investigated and re-think of it

**Comment by asdfjkl** [comment]
p 189 "a" missing: t FatFritz is just
Leela Chess Zero with the network trained in different way.

**Comment by asdfjkl** [comment]
p 82: clarify that black does not win when reaching the lowest rank, we only consider the outcomes white wins or black blocks white

**Comment by MSoszynski** [comment]
p9

Original: "It’s one thing to look in awe and admire the game changer AlphaZero, but it’s another one to figure out how AlphaZero and similar engines work and get a better understanding what they are capable of."

Better: "It’s one thing to look in awe and admire the game changer AlphaZero, it’s another thing to figure out how AlphaZero and similar engines work and get a better understanding of what they are capable."

**Comment by scottmcm** [comment]
v1.5, p37

> Indeed, if the network becomes moor accurate, our loss decreases.

should be "more accurate".

**Comment by asdfjkl** [comment]
p205... have think carefully should be have _to_ think carefully

**Comment by asdfjkl** [comment]
p204 ... on rather low-end office computer -> on _a_ rather ...

**Comment by asdfjkl** [comment]
p216: cite https://github.com/official-stockfish/Stockfish/pull/2916

**Comment by asdfjkl** [comment]
p232: Hexapawn is _a_ solved game,

**Comment by asdfjkl** [comment]
p219: the difference
is “only” 1.5 centipawns -> should be 1.5 pawns

**Comment by rodinhart** [comment]
Excellent book!

found in version 1.5:

page 24: I think it should be: w_0 = -10

page 31: I think the general rule of thumb for updating weights needs index i for the denominator:
  w_i^' = w_i = \nu dE_total / dw_**i**

**Comment by asdfjkl** [comment]
p8: DeepBlue should be Deep Blue

**Comment by octopus-prime** [comment]
What about page 198 "enemy king on e8, own knight on c3 from 1 to 0" ?!

I am not sure, but 
* there was no piece on c3
* the black king should see the white kinght on c3 too

So i would expect "enemy king on e8, own knight on c3 from 0 to 1"

**Comment by MetriC-DT** [comment]
Thanks for this great resource! Went through Ch. 2 (Back-Propagation and Gradient Descent), and I think there might be an issue with the partial derivatives (p. 30)

It should be (for $j = 1, 2$ in the example and $N = 2$):

$$
\frac{\partial E}{\partial w_j} = \sum_{i=1}^{N}\frac{\partial E}{\partial \text{out}_i} \cdot \frac{\partial \text{out}_i}{\partial \text{net}_i} \cdot \frac{\partial \text{net}_i}{\partial w_j} 
$$

The corresponding calculations might need to be re-worked as well.

**Comment by asdfjkl** [comment]
p62: There is one large and deep neural neural network. (neural twice)

**Comment by asdfjkl** [comment]
p172:formatting of nxf6 in bullet point (should use \mathrm)

**Comment by asdfjkl** [comment]
Figure 4.12: "convolution" on the left: font is too small

**Comment by asdfjkl** [comment]
(all fixed in version 1.6)

**Comment by MrOlegus** [comment]
Page 118. In the formula, Vi must be at the bottom and Vparent at the top. Is not it?

**Comment by cakiki** [comment]
Page 198: 

> "No book about computers is complete with really bad car analogies."

This should rather read: **without** really bad car analogies

**Comment by boihhs** [comment]
Page 39. The formula for the softmax function does not have an e on the denominator

**Comment by RickGroszkiewicz** [comment]
Page 16:

> "If we listen to Grandmaster's advice"

S/B "If we listen to Grandmasters' advice"


> "Look careful at the mistakes"

S/B "Look carefully at the mistakes"


Page 23:

> "and loose."

S/B "and lose."

**Comment by NelloHo** [comment]
p 39

![image](https://github.com/asdfjkl/neural_network_chess/assets/99171663/9d38ffd0-6896-4d68-bfe6-5c7eb5757922)

I guess the softmax function should be :thinking: 

![image](https://github.com/asdfjkl/neural_network_chess/assets/99171663/250cbf95-4e3c-40ad-ba21-6af93ca68407)

**Comment by NelloHo** [comment]
p. 85
![image](https://github.com/asdfjkl/neural_network_chess/assets/99171663/86a68565-fd92-4000-b996-cd4e8ca72f66)

I guess line 22 should be:
```python
minimax ( board , depth -1 , not minimize)
```
as the child nodes of a of a black node are white nodes, which should be maximized.

**Comment by NelloHo** [comment]
p. 123
![image](https://github.com/asdfjkl/neural_network_chess/assets/99171663/92d56613-2e15-4126-8ce3-adf16c68df83)
3. Qxc4 is not a legal move 🤔 , hence it wont be checked during MCTS

**Comment by RickGroszkiewicz** [comment]
Page 30:

In the chain rule for computing partial derivatives, the last term has delta-w-sub-i in the denominator.

S/B "delta-w-sub-1"

**Comment by marcps70** [comment]
several "loose" and "looses" that I think should be "lose" or "loses" on pages 23, 40, 96, 140, 210, 217

**Comment by tttony** [comment]
In page 200, here you meant: **Then**, right?
Thanks for the book

![20240308_125722](https://github.com/asdfjkl/neural_network_chess/assets/219885/18e5de8d-b54d-4421-b36c-96f45c042b50)

**Comment by xenos1984** [comment]
p.121 and mcts.py sample code:

In `simulate`, shouldn't payout be 0 if `OPPONENT` wins and 0.5 if nobody wins?

In `backpropagation`, the same payout is recorded at each level in the tree. In the MCTS algorithm, shouldn't the recorded payout at each level be the one for the player who is to move at that level? In the example code, a win (payout of 1) is recorded at each level when white wins, but that means that the selection step will *always* prefer moves in which white wins, even at those levels where black is to move, and thus it will actually prefer bad choices for black which are unlikely to be made.

**Comment by ppeloton** [comment]
First of all thanks a lot for writing this great book. 

A couple of comments regarding the section 4.6  about NNUEs
- HalfKP architecture. It would be good to clarify that this architecture uses rotational mirroring (as is shown in the figures 4.14 and 4.15) and not vertical which would be more intuitive for a chess player. 
- There is a new architecture HalfKA (I guess it was not there, when the book was written). If there is a new version of the book, it would be worth mentioning. There you also have vertical mirroring.
- p. 199 "Assume we always start at a1 and work towards h8". I think instead one should say that one always looks at the board from the side, whose turn it is to make a move, and you start from the bottom left corner to the upper right corner (in case of White from a1 to h8, in case of Black from h8 to a1). In the figures 4.14 and 4.15 you switched the naming of the squares, so your text is technically right. But I think it is still a bit misleading.

Besides it would make maybe sense to add some more remarks to the code. For example chapter 5, listing 5.18: It would make to remark that the `model`object is passed to the `learner` and when fitting the model to the new games played the `model`object inside the `learner`is updated as objects are passed by reference in Python. Again, this is probably a matter of taste and for some people self-evident, but I think it could help some readers.

**Comment by asdfjkl** [comment]
page 83: As we can see, the evaluation function give_s_ a
very rough assessment of the position

**Comment by actor10** [comment]
page 226/227:   no mention of black's 4th move.

**Comment by Peeda** [comment]
page 11: "Silman explains how to **characterizes** a position by imbalances, i.e. advantages and disadvantages that characterize the position" should be Silman explains how to characterize

**Comment by cakiki** [comment]
pages 100, 124 and 235 (all occurrences): "en-passent" needs to corrected to "en-passant".

**Comment by berniwal** [comment]
Very great book! Thank you very much for creating this.

I wondered why on page 118 you state the UCT formula with log(ChildVisits) at the top:

![Image](https://github.com/user-attachments/assets/56fb3cd7-d5e2-409b-ae23-54dfd0323674)

While I find mostly this definition online with log(ParentVisits) at the top:
https://en.wikipedia.org/wiki/Monte_Carlo_tree_search

![Image](https://github.com/user-attachments/assets/200b62b7-a369-4fa6-9272-0b796989d7b0)

**Comment by berniwal** [comment]
Thank you for this amazing book - I really enjoyed reading it.

I noticed that on page 118 you state the UCT with log(ChildVisits) at the top:

![Image](https://github.com/user-attachments/assets/b4d2528a-973d-4ffe-8f0e-af9cd9bc6213)

While if I search for the formula online I get log(ParentVisits) at the top:
https://en.wikipedia.org/wiki/Monte_Carlo_tree_search
https://www.chessprogramming.org/UCT

![Image](https://github.com/user-attachments/assets/c434a4d5-615b-4348-b476-c58c3938f4ac)

In your code example you also compute it like this:

![Image](https://github.com/user-attachments/assets/e09e56c9-f5d1-4880-a5cc-b570c0dc42fa)

**Comment by xianminx** [comment]
> p 39
> 
> ![image](https://github.com/asdfjkl/neural_network_chess/assets/99171663/9d38ffd0-6896-4d68-bfe6-5c7eb5757922)
> 
> I guess the softmax function should be 🤔
> 
> ![image](https://github.com/asdfjkl/neural_network_chess/assets/99171663/250cbf95-4e3c-40ad-ba21-6af93ca68407)

yes, if $o_i$ represents the raw score, then 

$$ \sigma(\mathbf{o}) = \frac{e^{o_i}}{\sum_{j=1}^n e^{o_j}} for i = 1,...n $$

And if $o_i=e^{z_i}$ where $z_i$ is the raw score and is the normalized prob, then, the formula should be 

$$ \sigma(\mathbf{o}) = \frac{o_i}{\sum_{j=1}^n o_j} for i = 1,...n $$

The current description is unclear. Could you suggest a clearer alternative?

**Comment by xianminx** [comment]
P75,P76, 

the first move has a ~~change~~ chance of 30 percent to be played.

**Comment by markhudson42** [comment]
There is the incorrect "Bakinman" instead of the correct "Baikinman" on page 62.

**Comment by Cubix1729** [comment]
p.97, "DeepBlues true capabilites" instead of "DeepBlue's true capabilities"

**Comment by Cubix1729** [comment]
p.107, "the most simplest" instead of "the most simple"
